Marshmallow

mgranberry opened the pull request:
Initial stab at marshmallow port.  Share support is completely broken by the retrofit update and permissions aren't being handled for export/import yet, but it is working better for doze (after turning off battery optimization).  

I still need to add a popup to direct users to toggle the setting.

Not ready for merge.

mgranberryto committed:
Port to Retrofit 2.0 because Apache http is no longer supported for API 23.

Remove obsolete REST api.

Comment out Share.

mgranberryto committed:
honor timeouts

mgranberryto committed:
M alarm support

mgranberry commented:
This also fixes the long-standing REST redelivery problems.

mgranberryto committed:
clean up alarms and add support for waking from doze where appropriate

mgranberry commented:
I'll need someone with some of the other devices to test this.  I still only use a Share 2 receiver with any regularity.

StephenBlackWasAlreadyTaken commented:
I actually just got the 5x (I havnt had it uploading or anything like that) I can definitely test them out, but my transmitter just died last night so now all I have is g5 transmitters
soooooo... I cant test them with the collection methods if thats a concern

mgranberry commented:
It looks like IntelliJ reformatted some whitespace.  Let me know if you want me to back out the whitespace changes.

jstevensog commented:
Any chance we could separate that fix into its own PR? Would love to have that fix now.
Cheers

---- Matthias Granberry wrote ----

> This also fixes the long-standing REST redelivery problems.
> 
> —
> Reply to this email directly or view it on GitHub.￼

mgranberry commented:
I think just ripping out all the try/catch blocks and adding "throws Exception" to all of the REST uploader methods would do it, but I ripped out almost all of the original code to do it.  It shouldn't be too hard.

jstevensog commented:
I am also looking at getting rid of the deprecated code elements, just because.  Not getting reliable connections to REST.  Very flakey.

---- Matthias Granberry wrote ----

> I think just ripping out all the try/catch blocks and adding "throws Exception" to all of the REST uploader methods would do it, but I ripped out almost all of the original code to do it. It shouldn't be too hard.
> 
> —
> Reply to this email directly or view it on GitHub.￼

mgranberry commented:
I haven't missed any dots today due to REST issues.  I'll work on getting the Share code working tonight so I can just merge everything, but that will either need testing or I'll have to provision myself using the in-laws' iPhones.

tzachi-dar commented:
I can test it, but I'm not using the rest api. Will that help?

Thanks
Tzachi

On Mon, Nov 2, 2015 at 11:48 PM, Matthias Granberry <
notifications@github.com> wrote:

> I haven't missed any dots today due to REST issues. I'll work on getting
> the Share code working tonight so I can just merge everything, but that
> will either need testing or I'll have to provision myself using the
> in-laws' iPhones.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/StephenBlackWasAlreadyTaken/xDrip-Experimental/pull/171#issuecomment-153167883
> .

mgranberryto committed:
request CONNECTION_PRIORITY_HIGH during Share2 transfers.  Drop back to CONNECTION_PRIORITY_LOW_POWER afterwards

mgranberryto committed:
clean up a typo

briancripe commented:
Nexus 6P user here.  Testing Dexcom G4 w/ Share transmitter, G4 sensor, and xDrip watchface on Pebble OG.  Local devices only, no sharing to cloud (for now).

Only have been running it for a little bit now, but so far collection service seems to be working good.

However, I had to add `android.permission.ACCESS_COARSE_LOCATION` to the Android Manifest and enable Location permission first, as [BLE Scanner now requires some form of location](http://stackoverflow.com/questions/33052811/since-marshmallow-update-bluetooth-discovery-using-bluetoothadapter-getdefaultad) to get the list of discovered devices to pair with.  Probably best to ask for permission at runtime in BluetoothScan activity onCreate().

Commit:
changed created_at to new ns specification for direct upload

mgranberryto committed:
use US locale for date formatting to enforce ASCII in date strings.

mgranberryto committed:
send multiple calibrations/BGs/raws in a single REST call

mgranberryto committed:
fix a potential crash if battery information isn't available

mgranberryto committed:
ShareRest ported

mgranberryto committed:
Location required for bluetooth scan in Marshmallow

mgranberryto committed:
Lay groundwork for session initialization code path

mgranberry commented:
@XenophonCydrome thanks.  I missed that.

mgranberry commented on app/src/main/java/com/eveningoutpost/dexdrip/FollowerManagementActivity.java:
```diff
@@ -12,25 +12,27 @@
-    FollowerManager followerManager;
-    Action1<List<ExistingFollower>> existingFollowerListener;
+    ShareRest shareRest;
+    Callback<List<ExistingFollower>> existingFollowerListener;
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Use explicit scoping instead of the default package private level](https://www.codacy.com/app/matthias-granberry/xDrip-Experimental/file/1296698328/issues/source?bid=2735053&fileBranchId=2947762#l35)

mgranberry commented on app/src/main/java/com/eveningoutpost/dexdrip/FollowerManagementActivity.java:
```diff
@@ -65,25 +67,32 @@ protected void onResume() {
             }
+
+            @Override
+            public void onFailure(Throwable t) {
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Document empty method body](https://www.codacy.com/app/matthias-granberry/xDrip-Experimental/file/1296698328/issues/source?bid=2735053&fileBranchId=2947762#l89)

mgranberry commented on app/src/main/java/com/eveningoutpost/dexdrip/UtilityModels/NightscoutUploader.java:
```diff
@@ -110,157 +133,73 @@ private boolean doRESTUpload(SharedPreferences prefs, List<BgReading> glucoseDat
-                            post.setEntity(se);
-                            post.setHeader("Accept", "application/json");
-                            post.setHeader("Content-type", "application/json");
+        private void doRESTUploadTo(NightscoutService nightscoutService, String secret, List<BgReading> glucoseDataSets, List<Calibration> meterRecords, List<Calibration> calRecords) throws Exception {
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [The method doRESTUploadTo() has an NPath complexity of 16](https://www.codacy.com/app/matthias-granberry/xDrip-Experimental/file/1296698421/issues/source?bid=2735053&fileBranchId=2947762#l178)

mgranberry replied:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Avoid long parameter lists.](https://www.codacy.com/app/matthias-granberry/xDrip-Experimental/file/1296698421/issues/source?bid=2735053&fileBranchId=2947762#l178)

mgranberry replied:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [The method doRESTUploadTo() has an NPath complexity of 16](https://www.codacy.com/app/matthias-granberry/xDrip-Experimental/file/1305335990/issues/source?bid=2735053&fileBranchId=2947762#l178)

mgranberry replied:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Avoid long parameter lists.](https://www.codacy.com/app/matthias-granberry/xDrip-Experimental/file/1305335990/issues/source?bid=2735053&fileBranchId=2947762#l178)

mgranberry replied:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [The method doRESTUploadTo() has an NPath complexity of 16](https://www.codacy.com/app/matthias-granberry/xDrip-Experimental/file/1306138886/issues/source?bid=2735053&fileBranchId=2947762#l178)

mgranberry replied:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Avoid long parameter lists.](https://www.codacy.com/app/matthias-granberry/xDrip-Experimental/file/1306138886/issues/source?bid=2735053&fileBranchId=2947762#l178)

mgranberry commented on app/src/main/java/com/eveningoutpost/dexdrip/ShareModels/ShareRest.java:
```diff
@@ -15,85 +30,64 @@
-                adapter.create(DexcomShareInterface.class);
-        return checkSessionActive;
-    }
+    SharedPreferences.OnSharedPreferenceChangeListener preferenceChangeListener = new SharedPreferences.OnSharedPreferenceChangeListener() {
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Use explicit scoping instead of the default package private level](https://www.codacy.com/app/matthias-granberry/xDrip-Experimental/file/1296698395/issues/source?bid=2735053&fileBranchId=2947762#l51)

mgranberry commented on app/src/main/java/com/eveningoutpost/dexdrip/UtilityModels/NightscoutUploader.java:
```diff
@@ -55,24 +59,43 @@
+        }
+
+        private class UploaderException extends RuntimeException {
+            int code;
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Use explicit scoping instead of the default package private level](https://www.codacy.com/app/matthias-granberry/xDrip-Experimental/file/1296698421/issues/source?bid=2735053&fileBranchId=2947762#l80)

mgranberry replied:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Use explicit scoping instead of the default package private level](https://www.codacy.com/app/matthias-granberry/xDrip-Experimental/file/1305335990/issues/source?bid=2735053&fileBranchId=2947762#l80)

mgranberry replied:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Use explicit scoping instead of the default package private level](https://www.codacy.com/app/matthias-granberry/xDrip-Experimental/file/1306138886/issues/source?bid=2735053&fileBranchId=2947762#l80)

mgranberry commented on app/src/main/java/com/eveningoutpost/dexdrip/ShareModels/ShareRest.java:
```diff
@@ -134,15 +156,112 @@ public boolean verify(String hostname, SSLSession session) {
+        public AuthenticatingCallback (Callback<T> callback) {
+            this.delegate = callback;
+        }
+
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Use explicit scoping instead of the default package private level](https://www.codacy.com/app/matthias-granberry/xDrip-Experimental/file/1296698395/issues/source?bid=2735053&fileBranchId=2947762#l235)

mgranberry commented on app/src/main/java/com/eveningoutpost/dexdrip/Services/SyncService.java:
```diff
@@ -3,26 +3,19 @@
 import com.eveningoutpost.dexdrip.UtilityModels.MongoSendTask;
-import com.eveningoutpost.dexdrip.UtilityModels.NightscoutUploader;
-import com.eveningoutpost.dexdrip.UtilityModels.RestCalls;
 import com.eveningoutpost.dexdrip.UtilityModels.SensorSendQueue;
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Avoid unused imports such as 'com.eveningoutpost.dexdrip.UtilityModels.SensorSendQueue'](https://www.codacy.com/app/matthias-granberry/xDrip-Experimental/file/1296698384/issues/source?bid=2735053&fileBranchId=2947762#l16)

mgranberry replied:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Avoid unused imports such as 'com.eveningoutpost.dexdrip.UtilityModels.SensorSendQueue'](https://www.codacy.com/app/matthias-granberry/xDrip-Experimental/file/1305335953/issues/source?bid=2735053&fileBranchId=2947762#l16)

mgranberry replied:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Avoid unused imports such as 'com.eveningoutpost.dexdrip.UtilityModels.SensorSendQueue'](https://www.codacy.com/app/matthias-granberry/xDrip-Experimental/file/1306138849/issues/source?bid=2735053&fileBranchId=2947762#l16)

mgranberry commented on app/src/main/java/com/eveningoutpost/dexdrip/UtilityModels/NightscoutUploader.java:
```diff
@@ -110,157 +133,73 @@ private boolean doRESTUpload(SharedPreferences prefs, List<BgReading> glucoseDat
+                    String secret = uri.getUserInfo();
+                    if ((secret == null || secret.isEmpty()) && apiVersion == 0) {
+                        baseURL = baseURI;
+                    } else if ((secret == null || secret.isEmpty())) {
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Useless parentheses.](https://www.codacy.com/app/matthias-granberry/xDrip-Experimental/file/1296698421/issues/source?bid=2735053&fileBranchId=2947762#l143)

mgranberry replied:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Useless parentheses.](https://www.codacy.com/app/matthias-granberry/xDrip-Experimental/file/1305335990/issues/source?bid=2735053&fileBranchId=2947762#l143)

mgranberry replied:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Useless parentheses.](https://www.codacy.com/app/matthias-granberry/xDrip-Experimental/file/1306138886/issues/source?bid=2735053&fileBranchId=2947762#l143)

mgranberry commented on app/src/main/java/com/eveningoutpost/dexdrip/ShareModels/ShareRest.java:
```diff
@@ -134,15 +156,112 @@ public boolean verify(String hostname, SSLSession session) {
+        //            checkMonitorAssignment
+        //              !AssignedToYou: updateMonitorAssignment
+        //                  getValidSessionId
+
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Avoid empty if statements](https://www.codacy.com/app/matthias-granberry/xDrip-Experimental/file/1296698395/issues/source?bid=2735053&fileBranchId=2947762#l176)

mgranberryto committed:
Request filesystem permissions

mgranberryto committed:
remove some unused state from class

mgranberry commented:
I think this is ready to go now.

mgranberryto committed:
Register new receivers as well.

Clean up some unused imports, unused variables.

mgranberry commented on app/src/main/java/com/eveningoutpost/dexdrip/ShareModels/ShareRest.java:
```diff
@@ -130,19 +153,133 @@ public boolean verify(String hostname, SSLSession session) {
+                sessionId = dexcomShareApi.getSessionId(new ShareAuthenticationBody(password, username).toMap()).execute().body();
+                dexcomShareApi.authenticatePublisherAccount(sessionId, serialNumber, new ShareAuthenticationBody(password, username).toMap()).execute().body();
+                dexcomShareApi.StartRemoteMonitoringSession(sessionId, serialNumber).execute();
+                String assignment = dexcomShareApi.checkMonitorAssignment(sessionId, serialNumber).execute().body();
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Position literals first in String comparisons](https://www.codacy.com/app/matthias-granberry/xDrip-Experimental/file/1305335964/issues/source?bid=2735053&fileBranchId=2947762#l183)

mgranberry replied:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Position literals first in String comparisons](https://www.codacy.com/app/matthias-granberry/xDrip-Experimental/file/1306138860/issues/source?bid=2735053&fileBranchId=2947762#l183)

mgranberry commented on app/src/main/java/com/eveningoutpost/dexdrip/Services/SyncService.java:
```diff
@@ -3,26 +3,19 @@
-import com.eveningoutpost.dexdrip.Models.BgReading;
-import com.eveningoutpost.dexdrip.Models.Calibration;
+import com.eveningoutpost.dexdrip.Models.UserError.Log;
 import com.eveningoutpost.dexdrip.UtilityModels.BgSendQueue;
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Avoid unused imports such as 'com.eveningoutpost.dexdrip.UtilityModels.BgSendQueue'](https://www.codacy.com/app/matthias-granberry/xDrip-Experimental/file/1305335953/issues/source?bid=2735053&fileBranchId=2947762#l13)

mgranberry replied:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Avoid unused imports such as 'com.eveningoutpost.dexdrip.UtilityModels.BgSendQueue'](https://www.codacy.com/app/matthias-granberry/xDrip-Experimental/file/1306138849/issues/source?bid=2735053&fileBranchId=2947762#l13)

mgranberry commented on app/src/main/java/com/eveningoutpost/dexdrip/ShareModels/DexcomShare.java:
```diff
@@ -0,0 +1,133 @@
+    // returns true or false
+
+    @POST("Publisher/StartRemoteMonitoringSession")
+    Call<ResponseBody> StartRemoteMonitoringSession(@Query("sessionId") String sessionId,
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Method names should not start with capital letters](https://www.codacy.com/app/matthias-granberry/xDrip-Experimental/file/1305335957/issues/source?bid=2735053&fileBranchId=2947762#l34)

mgranberry replied:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Method names should not start with capital letters](https://www.codacy.com/app/matthias-granberry/xDrip-Experimental/file/1306138853/issues/source?bid=2735053&fileBranchId=2947762#l34)

mgranberry commented on app/src/main/java/com/eveningoutpost/dexdrip/ShareModels/DexcomShare.java:
```diff
@@ -0,0 +1,133 @@
+package com.eveningoutpost.dexdrip.ShareModels;
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Package name contains upper case characters](https://www.codacy.com/app/matthias-granberry/xDrip-Experimental/file/1305335957/issues/source?bid=2735053&fileBranchId=2947762#l1)

mgranberryto committed:
avoid crash

mgranberryto committed:
request battery that a user disables battery optimization during doze.  I think we're justified.

mgranberry commented:
I think this is complete, but it is a pretty big change.  I have testing the most-exercised changes for a week, but the NS REST uploader and Share uploader are almost completely rewritten and doze support is only lightly tested.  I've been using most of it for a week and it has been significantly better than before, but everyone's device behaves differently.

I'm afraid that I may need to remove raw syncing from the Share2 source when dozing in order to guarantee staying under the 10-second execution period.

mgranberryto committed:
widget doesn't need to update when device is idle.  Don't wake the device.

mgranberryto committed:
Allow SyncService to piggyback on collector wakes.

mgranberryto committed:
Use alarms as a last resort and wake based on "early" heartbeats instead to give a little extra protection when dozing.

mgranberryto committed:
don't miss retries when the heartbeat timer is nearly equal to a minute

mgranberryto committed:
Use AlarmManager.setAllarmClock() call to enable alarms during doze.

mgranberry commented:
I fixed the low/high alarms during Doze and made the Dexcom Share2 collector much more robust during doze, only requiring alarms when the device becomes disconnected.  I was able to run my phone on Doze for 4h and didn't observe any issues.  Wifi wixels are the only questionable device, but this should only improve things for them.

mgranberryto committed:
Waste less battery when waiting on alarms.

mgranberryto committed:
Don't wake to update widget when screen is off.  Update widget when screen turns on.

mgranberryto committed:
Copy-paste error, fix SyncService alarm

mgranberryto committed:
be less aggressive about fetching data

mgranberryto committed:
don't wake in SyncService unnecessarily.

mgranberryto committed:
rename widgetUpdateService to WidgetUpdateService, wait for next data tick to upload data, and activate the alarm only if the next data point gets missed

mgranberry commented:
This completes my war on device wakes.  It might be possible to do some more for the xDrips, but I don't feel comfortable optimizing too much until I get around to assembling mine.

jstevensog commented:
Once I get Marshmallow on my Z3, I will be more than happy to assist with
testing xBridge with it.
Thanks for all this work Matthias.  Exceptional.
Cheers

On Fri, Nov 13, 2015 at 9:07 AM, Matthias Granberry <
notifications@github.com> wrote:

> This completes my war on device wakes. It might be possible to do some
> more for the xDrips, but I don't feel comfortable optimizing too much until
> I get around to assembling mine.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/StephenBlackWasAlreadyTaken/xDrip-Experimental/pull/171#issuecomment-156251110
> .

## 

John Stevens
"You are how you live, not what you have."

tzachi-dar commented:
I'm running this branch on android 4.42 with wifi wixel. best I can do
right now.
so far with no issues, hope to get a newer device next week.

HTH
Tzachi

On Fri, Nov 13, 2015 at 12:22 AM, John notifications@github.com wrote:

> Once I get Marshmallow on my Z3, I will be more than happy to assist with
> testing xBridge with it.
> Thanks for all this work Matthias. Exceptional.
> Cheers
> 
> On Fri, Nov 13, 2015 at 9:07 AM, Matthias Granberry <
> notifications@github.com> wrote:
> 
> > This completes my war on device wakes. It might be possible to do some
> > more for the xDrips, but I don't feel comfortable optimizing too much
> > until
> > I get around to assembling mine.
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > <
> > https://github.com/StephenBlackWasAlreadyTaken/xDrip-Experimental/pull/171#issuecomment-156251110
> > 
> > .
> 
> ## 
> 
> John Stevens
> "You are how you live, not what you have."
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/StephenBlackWasAlreadyTaken/xDrip-Experimental/pull/171#issuecomment-156255446
> .

Commit:
add disable battery warning to preferences

Commit:
disable battery warning

AdrianLxM commented on app/src/main/java/com/eveningoutpost/dexdrip/UtilityModels/Notifications.java:
```diff
@@ -253,37 +254,57 @@ public void notificationSetter(Context context) {
-                }
+                // sleep longer if the alert is snoozed.
+                long wakeTime = activeBgAlert.is_snoozed ? activeBgAlert.next_alert_at :
+                        calendar.getTimeInMillis() + (time + 60000);
```
Shouldn't this be `... + (time * 60000);`?

Commit:
wakelock + time

Added a wakelock to onHandleIntent.
The few more instructions on `ArmTimer()` were enough for my phone to decide to sleep.

Exchanged `+` for an `*` operator.

mgranberry committed:
Merge pull request #176 from StephenBlackWasAlreadyTaken/marshmallow-patches01

wakelock + time

mgranberryto committed:
Merge branch 'master' into marshmallow

mgranberry committed:
Merge pull request #175 from StephenBlackWasAlreadyTaken/marshmallow-batterywarning

Marshmellow batterywarning

Commit:
small text change

AdrianLxM commented:
Another Problem I've found is: The ascending alerts rely on the timer being called every minute.
With rerise periods longer than one minute it won't work.

As this is reraise related I'll fix this in the according branch. I guess without rewriting the AlarmSystem we just have to have "reraise" set to 1 minute if ascending alerts are enabled. (And tell the user that ascending alerts disable reraise)

mgranberry commented:
My intention was to have this code handle waking the device during an ascending alert:

```
                Calendar calendar = Calendar.getInstance();
                AlarmManager alarm = (AlarmManager) getSystemService(ALARM_SERVICE);
                // sleep longer if the alert is snoozed.
                long wakeTime = activeBgAlert.is_snoozed ? activeBgAlert.next_alert_at :
                        calendar.getTimeInMillis() + (time * 60000);
                Log.d(TAG , "ArmTimer waking at: "+ wakeTime);
                if (wakeIntent != null)
                    alarm.cancel(wakeIntent);
                wakeIntent = PendingIntent.getService(this, 0, new Intent(this, this.getClass()), 0);
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                    alarm.setAlarmClock(new AlarmManager.AlarmClockInfo(wakeTime, wakeIntent), wakeIntent);
                } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
                    alarm.setExact(AlarmManager.RTC_WAKEUP, wakeTime, wakeIntent);
                } else
                    alarm.set(AlarmManager.RTC_WAKEUP, wakeTime, wakeIntent);
```

It worked when I tried it, but maybe the rerise changes broke that?  Is is possible to add a method like `getNextWakeTime` that will handle the corner cases?  My intuition is that it should be the minimum of (`wakeTime`, `ascendingAlertTime`) when there is an unacknowledged alert out, but I thought that's what `next_alert_at` was supposed to do?  Clearly I need to read some code.

AdrianLxM commented:
Well, `time` depends on the rerise interval. If it is set to more than one minute, the second "louder" alert won't show.

`next_alert_at` seems to be intended for when the snooze period is over:

``` java
    public static void ClearIfSnoozeFinished() {
        Log.d(TAG, "ActiveBgAlert ClearIfSnoozeFinished called");
        ActiveBgAlert aba = getOnly();
        if (aba != null) {
            if(new Date().getTime() > aba.next_alert_at) {
                Log.d(TAG, "ActiveBgAlert ClearIfSnoozeFinished deleting allert");
                aba.delete();
            }
        }
    }
```

It seems to be used to determine if the snooze time is over. But on the other hand it is just called under certain circumstances and doesn't do what the name suggests. It is more a "delete active alert but keep Snooze active". At least that's what the comments in the control flow leading to this call suggest ("// No alert should work, Stop all alerts, but keep the snoozing...").

I honestly just want to be over with patching this code. Non-Share users do need the alarm system :/.

mgranberry commented:
Of course.  I think I'll revert to the original alarm behavior and use
setAlarmClock.

On Wed, Nov 18, 2015, 3:30 PM AdrianLxM notifications@github.com wrote:

> Well, time depends on the rerise interval. If it is set to more than one
> minute, the second "louder" alert won't show.
> 
> next_alert_at seems to be intended for when the snooze period is over:
> 
> ```
> public static void ClearIfSnoozeFinished() {
>     Log.d(TAG, "ActiveBgAlert ClearIfSnoozeFinished called");
>     ActiveBgAlert aba = getOnly();
>     if (aba != null) {
>         if(new Date().getTime() > aba.next_alert_at) {
>             Log.d(TAG, "ActiveBgAlert ClearIfSnoozeFinished deleting allert");
>             aba.delete();
>         }
>     }
> }
> ```
> 
> It seems to be used to determine if the snooze time is over. But on the
> other hand it is just called under certain circumstances and doesn't do
> what the name suggests. It is more a "delete active alert but keep Snooze
> active". At least that's what the comments in the control flow leading to
> this call suggest ("// No alert should work, Stop all alerts, but keep the
> snoozing...").
> 
> I honestly just want to be over with patching this code. Non-Share users
> do need the alarm system :/.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/StephenBlackWasAlreadyTaken/xDrip-Experimental/pull/171#issuecomment-157870485
> .

AdrianLxM commented:
By changing the interval to one minute in case "ascending" is activated it seems to work now.
As far as I remember it had the same problem before as well due to another check deeper in the system. If it now works, I would opt for better battery :)

mgranberryto committed:
Log wake time message in human-readable time

Commit:
removed global variable with just local intent of usage (caused side effect using it globally)

Conflicts:
	app/src/main/java/com/eveningoutpost/dexdrip/UtilityModels/AlertPlayer.java

tzachi-dar committed:
Menu changes: ascending disables alerts re-raise.

Signed-off-by: Tzachi Dar <tzachi.dar@gmail.com>

tzachi-dar committed:
Fix issue with re-rise not working correctly.

Signed-off-by: Tzachi Dar <tzachi.dar@gmail.com>
(still needs more testing)

tzachi-dar committed:
Fix a bug in alarm system.

mgranberry committed:
Merge pull request #183 from StephenBlackWasAlreadyTaken/marshmallow-rerase2

merge in Marshmallow rerase2 to have it get as much testing as possible

mgranberryto committed:
Merge remote-tracking branch 'origin/master' into marshmallow
